lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/TaskTrainingMethods.scala

```diff
@@ -0,0 +1,164 @@
+                    validationData: Option[Broadcast[Array[Row]]], log: Logger,
+                    trainParams: TrainParams, numTasksPerExec: Int, schema: StructType,
+                    sharedState: SharedState)
+                   (inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
+    var isEnabledWorker = false
+    val useSingleDatasetMode = trainParams.executionParams.useSingleDatasetMode
+    try {
+      val emptyPartition = !inputRows.hasNext
+      isEnabledWorker = if (!emptyPartition) isWorkerEnabled(trainParams, log, sharedState) else false
+      // Initialize the native library
+      LightGBMUtils.initializeNativeLibrary()
+      // Initialize the network communication
+      val (nodes, localListenPort) = getNetworkInfo(networkParams, numTasksPerExec, log, isEnabledWorker)
+      if (emptyPartition) {
+        log.warn("LightGBM task encountered empty partition, for best performance ensure no partitions empty")
+        List[LightGBMBooster]().toIterator
+      } else {
+        if (isEnabledWorker) {
+          log.info(s"LightGBM task listening on: $localListenPort")
+          if (useSingleDatasetMode) sharedState.helperStartSignal.countDown()
+        } else {
+          sharedState.helperStartSignal.await()
+        }
+        val (aggregatedColumns, aggregatedValidationColumns) = prepareDatasets(inputRows,
+          validationData, columnParams, schema, trainParams, sharedState)
+        // Return booster only from main worker to reduce network communication overhead
+        val returnBooster = getReturnBooster(isEnabledWorker, nodes, log, numTasksPerExec, localListenPort)
+        try {
+          if (isEnabledWorker) {
+            // If worker enabled, initialize the network ring of communication
+            networkInit(nodes, localListenPort, log, LightGBMConstants.NetworkRetries, LightGBMConstants.InitialDelay)
+            if (useSingleDatasetMode) sharedState.doneSignal.await()
+            translate(batchIndex, columnParams, aggregatedValidationColumns, log, trainParams, returnBooster,
+              sharedState.isSparse.get, schema, aggregatedColumns)
+          } else {
+            log.info("Helper task finished processing rows")
+            sharedState.doneSignal.countDown()
+            List[LightGBMBooster]().toIterator
+          }
+        } finally {
+          // Finalize network when done
+          if (isEnabledWorker) LightGBMUtils.validate(lightgbmlib.LGBM_NetworkFree(), "Finalize network")
+        }
+      }
+    } finally {
+      resetState(useSingleDatasetMode, isEnabledWorker, log, sharedState)
+    }
+  }
+
+  def translate(batchIndex: Int, columnParams: ColumnParams, validationData: Option[AggregatedColumns],
+                log: Logger, trainParams: TrainParams, returnBooster: Boolean, isSparse: Boolean, schema: StructType,
+                aggregatedColumns: AggregatedColumns): Iterator[LightGBMBooster] = {
+    var trainDatasetOpt: Option[LightGBMDataset] = None
```

Comment:
Seems these vars could be vals?

Reply:
they are vars so that they can be removed in the finally block:
```
    } finally {
      // Free datasets
      trainDatasetOpt.foreach(_.close())
      validDatasetOpt.foreach(_.close())
    }
```
do you see a better way to make them vals but ensure they get released in finally block?